src/swift/Sources/Models/TokenFile.swift

```diff
@@ -168,59 +183,49 @@ class TokenFile: Codable {
         text(packageName)
         whitespace()
         punctuation("{")
+        newLine()
         indent {
             declarations.forEach { topLevelDecl in
                 process(topLevelDecl)
             }
         }
-        if needsNewLine {
-            newLine()
-        }
         punctuation("}")
         newLine()
 
-        navigation(from: declarations)
+        navigationTokens(from: declarations)
         // ensure items appear in sorted order
-        navigation.sort(by: {$0.text < $1.text })
-        for item in navigation {
+        navigation.sort(by: { $0.text < $1.text })
+        navigation.forEach { item in
             item.childItems.sort(by: { $0.text < $1.text })
         }
     }
 
-    private func process(_ decl: TopLevelDeclaration) {
-        if needsNewLine {
-            newLine()
-        }
-        for statement in decl.statements {
+    private func process(_ decl: TopLevelDeclaration, overridingAccess: AccessLevelModifier? = nil) {
+        let stopIdx = decl.statements.count - 1
+        for (idx, statement) in decl.statements.enumerated() {
             switch statement {
             case let decl as Declaration:
-                process(decl)
+                if process(decl, overridingAccess: overridingAccess), idx != stopIdx {
+                    // add an blank line for each declaration that is actually rendered
+                    newLine()
+                }
             default:
-                continue
+                SharedLogger.fail("Unsupported statement type: \(statement)")
             }
         }
     }
 
-    private func process(_ decl: ClassDeclaration) {
-        // Gather Information
-        SharedLogger.debug(decl.textDescription)
-        SharedLogger.debug(decl.attributes.textDescription)
-        SharedLogger.debug(decl.genericWhereClause?.textDescription ?? "")
+    private func process(_ decl: ClassDeclaration, overridingAccess: AccessLevelModifier? = nil) -> Bool {
```

Comment:
I feel that if similar declarations were grouped with protocols we could do even more refactoring. A lot of the implementation doesn't seem declaration-specific.

Reply:
It may be possible to refactor class/struct (and later actor) into a common rendering method like I did for `processMember`, where the only difference is essentially using `let` or `var`. I'll take a look.

Reply:
I attempted to do this but it doesn't work because of the way `ClassDeclaration.Member` and `StructDeclaration.Member` are defined by `swift-ast`. While these types both have a `member` collection of type `[Member]` the types are not compatible and there's no protocol that links them. Us attempting to define and conform third-party types to our protocols will likely just result in moving things around rather than becoming more concise.